File: hacinamiento_microdatos.R
##Borrar datos del entorno
rm(list=ls())


#Directorio de trabajo
setwd("D:/sedatu/rezago/")

# Librerías ====
if(!require('pacman')) install.packages('pacman')
pacman::p_load(tidyverse, srvyr,
               kableExtra)

#Descarga de archivos
url<-"https://www.inegi.org.mx/contenidos/programas/enigh/nc/2018/microdatos/enigh2018_ns_viviendas_csv.zip"

##Creación de directorio temporal

td<- tempdir()

# Descarga del archivo temporal


tf = tempfile(tmpdir=td, fileext=".zip")
download.file(url, tf)

# unzip

unzip(tf, files="viviendas.csv", exdir=td, 
      overwrite=TRUE)
fpath=file.path(td,"viviendas.csv")
unlink(td)

#Leer el archivo

enigh<-read.csv(fpath)%>%
  #Renombrar folioviv
  rename(folioviv=1)%>%
  #Obtener la clave de entidad de la variable folioviv
  mutate(cve_ent=
           case_when(nchar(folioviv)==9 ~ substr(folioviv,1,1),
                     TRUE ~substr(folioviv,1,2)),
#Creamos nombre de entidad
                 nom_ent= case_when(cve_ent==1 ~ "Aguascalientes",
                       cve_ent==2 ~ "Baja California",
                       cve_ent==3 ~ "Baja California Sur",
                       cve_ent==4 ~ "Campeche",
                       cve_ent==5 ~ "Coahuila de Zaragoza",
                       cve_ent==6 ~ "Colima",
                       cve_ent==7 ~ "Chiapas",
                       cve_ent==8 ~ "Chihuahua",
                       cve_ent==9 ~ "Ciudad de México",
                       cve_ent==10 ~ "Durango",
                       cve_ent==11 ~ "Guanajuato",
                       cve_ent==12 ~ "Guerrero",
                       cve_ent==13 ~ "Hidalgo",
                       cve_ent==14 ~ "Jalisco",
                       cve_ent==15 ~ "México",
                       cve_ent==16 ~ "Michoacán de Ocampo",
                       cve_ent==17 ~ "Morelos",
                       cve_ent==18 ~ "Nayarit",
                       cve_ent==19 ~ "Nuevo León",
                       cve_ent==20 ~ "Oaxaca",
                       cve_ent==21 ~ "Puebla",
                       cve_ent==22 ~ "Querétaro",
                       cve_ent==23 ~ "Quintana Roo",
                       cve_ent==24 ~ "San Luis Potosí",
                       cve_ent==25 ~ "Sinaloa",
                       cve_ent==26 ~ "Sonora",
                       cve_ent==27 ~ "Tabasco",
                       cve_ent==28 ~ "Tamaulipas",
                       cve_ent==29 ~ "Tlaxcala",
                       cve_ent==30 ~ "Veracruz de Ignacio de la Llave",
                       cve_ent==31 ~ "Yucatán",
                       cve_ent==32 ~ "Zacatecas"),
  #Creamos indicador de hacinamiento
  hac=case_when((tot_resid / num_cuarto) > 2.5 ~ 1,
                TRUE ~ 0))


#Definir diseño muestral

mydesign <-enigh%>%
  as_survey_design(ids=upm,
                   strata=est_dis,
                   weights=factor)

#Prueba de significancia estadística


#Viviendas con habitantes en condición de hacinamiento nacional
hacnal<-mydesign %>%
  filter(hac==1)%>%
  summarise(hac=survey_total(
    vartype = c("cv", "ci"),
    level=0.95))%>%
  mutate(hac_cv=
           hac_cv*100,
#Formato de números
hac=format(hac,big.mark=","),
hac_cv=round(hac_cv,2),
hac_low=format(hac_low,big.mark=","),
hac_upp=format(hac_upp,big.mark=","),
)

#Tabla Nacional
hacnal%>%
kable(caption=text_spec("México. Viviendas con habitantes en condición de hacinamiento, 2018",
                        bold=T, color="black",font_size = 30),
      format="html",
      align = "c",
      col.names = c("Viviendas con  habitantes en condición de\n
                    hacinamiento",
                    "Coeficiente de variación",
                    "Límite inferior",
                    "Límite superior"))%>%
  kable_styling(full_width = F, font_size = 20,
                html_font = "Montserrat Medium")%>%
  row_spec(0, bold = F, color = "black", background = "#9ebcda")%>%
  footnote(general = "@claudiodanielpc con información de INEGI. Encuesta Nacional de Ingresos y Gastos de los Hogares (ENIGH) 2018",
           general_title = "
Fuente: ")%>%
  #Salvar
  as_image(file="hacnal.png")


#Viviendas con habitantes en condición de hacinamiento por entidad federativa
hacent<-mydesign %>%
  filter(hac==1)%>%
  group_by(nom_ent)%>%
  summarise(hac=survey_total(
    vartype = c("cv", "ci"),
    level=0.95))%>%
  #Multiplicar coeficiente de variación por 100
  mutate(hac_cv=
           hac_cv*100,
         #Formato de números
         hac=format(hac,big.mark=","),
         hac_cv=round(hac_cv,2),
         hac_low=format(hac_low,big.mark=","),
         hac_upp=format(hac_upp,big.mark=","),
  )

#Tabla por entidad federativa
hacent%>%
  kable(caption=text_spec("Viviendas con habitantes en condición de hacinamiento por entidad federativa, 2018",
                          bold=T, color="black",font_size = 30),
        format="html",
        align = "c",
        col.names = c("Entidad",
        "Viviendas con  habitantes 
en condición de\n
                    hacinamiento",
                      "Coeficiente de variación",
                      "Límite inferior",
                      "Límite superior"))%>%
  kable_styling(full_width = F, font_size = 20,
                html_font = "Montserrat Medium")%>%
  row_spec(0, bold = F, color = "black", background = "#9ebcda")%>%
  footnote(general = "@claudiodanielpc con información de INEGI. Encuesta Nacional de Ingresos y Gastos de los Hogares (ENIGH) 2018",
           general_title = "
Fuente: ")%>%
  #Salvar
  as_image(file="hacent.png")

File: hacinamiento_microdatos.py
import pandas as pd
from samplics.estimation import TaylorEstimator
from urllib.request import urlopen
from zipfile import ZipFile
from io import BytesIO
import plotly.graph_objects as go
import plotly.io as pio

#Descarga de archivos
url = "https://www.inegi.org.mx/contenidos/programas/enigh/nc/2018/microdatos/enigh2018_ns_viviendas_csv.zip"

##Descarga y descomprimir archivos
resp = urlopen(url)
zipfile = ZipFile(BytesIO(resp.read()))
zf.extractall()

#Lectura y selección de variables
df =pd.read_csv("viviendas.csv",
                usecols=["folioviv",
                         "num_cuarto",
                         "tot_resid",
                         "est_dis",
                         "upm",
                         "factor"])

#Generar indicador de hacinamiento
df["hac"] = ((df['tot_resid']/df["num_cuarto"])> 2.5).astype(int)


#Precisión en la estimación del hacinamiento

hac_prec = TaylorEstimator("total")
hac_prec.estimate(
    y=df["hac"],
    samp_weight=df["factor"],
    stratum=df["est_dis"],
    psu=df["upm"],
    remove_nan=True,
)

print(hac_prec)





#Transformar el objeto a string y dataframe para visualizar

#Se transforma a objeto de texto
texto=str(hac_prec)
#Reemplazamos espacios por tabs
texto=texto.replace(' ', '|')

#Transformamos a dataframe
prec = pd.DataFrame([x.split('|') for x in
                     texto.split('\n')])

#Se elimina la primera fila y se dejan las primeras cinco columnas
prec = prec.iloc[1:, 0:5]

#Renombrar columnas
columnas = {prec.columns[0]: 'tot',
            prec.columns[1]: 'se',
            prec.columns[2]: "low",
            prec.columns[3]: "up",
            prec.columns[4]: "cv"}

prec = prec.rename(columns=columnas)



#Formato de números
cols=prec.columns
prec[cols] = prec[cols].apply(pd.to_numeric,
                              errors='coerce')
#Coeficiente de variación como porcentaje
prec["cv"]=\
    prec["cv"]*100

format_dict = {
                "tot": "{0:,.0f}",
                "se": "{0:,.0f}",
                "low": "{0:,.0f}",
                "up": "{0:,.0f}",
               "cv": "{:.2f}"}

for key, value in format_dict.items():
    prec[key] = prec[key].apply(value.format)


#Tabla usando plotly
fig = go.Figure(data=[go.Table(
    header=dict(values=["<b>Viviendas con habitantes en condición de hacinamiento</b>",
                        "<b>Error estándar</b>",
                "<b>Límite inferior</b>",
                        "<b>Límite superior</b>",
                        "<b>Coeficiente de variación</b>"
                        ],
                fill_color='#9ebcda',
                align='center',
                font_family="Montserrat",
                font_size=20),
cells=dict(values=[prec["tot"], prec["se"], prec["low"],
                   prec["up"], prec["cv"]],
               fill_color='lavender',
               align='center',
           font_family="Montserrat",
           font_size=18))
])

#Título
fig.update_layout(
    showlegend=False,
    title_text="<b>México. Viviendas con habitantes en condición de hacinamiento, 2018<br>",
    title_font_family="Montserrat",
    title_font_size=25
)

#Fuente
fig.add_annotation(text="@claudiodanielpc con información de INEGI. Encuesta Nacional de Ingresos y Gastos de los Hogares (ENIGH) 2018",
                  x=0, y=0, showarrow=False,
                   font_family="Century Gothic")

#Salvar la tabla como PNG

pio.write_image(fig, "tablaenigh.png",width=2000, scale=1,
                height=400)